src/analysis/processing/qgsalgorithmmultiringbuffer.cpp

```diff
@@ -0,0 +1,139 @@
+{
+  return QObject::tr( "This algorithm computes multi-ring ('donuts') buffer for all the features in an input layer, using a fixed or dynamic distance and rings number." );
+}
+
+QgsMultiRingBufferAlgorithm *QgsMultiRingBufferAlgorithm::createInstance() const
+{
+  return new QgsMultiRingBufferAlgorithm();
+}
+
+void QgsMultiRingBufferAlgorithm::initParameters( const QVariantMap & )
+{
+  std::unique_ptr< QgsProcessingParameterNumber> rings = qgis::make_unique< QgsProcessingParameterNumber >( QStringLiteral( "RINGS" ),
+      QObject::tr( "Number of rings" ), QgsProcessingParameterNumber::Integer,
+      1, false, 0 );
+  rings->setIsDynamic( true );
+  rings->setDynamicPropertyDefinition( QgsPropertyDefinition( QStringLiteral( "RINGS" ), QObject::tr( "Number of rings" ), QgsPropertyDefinition::IntegerPositive ) );
+  rings->setDynamicLayerParameterName( QStringLiteral( "INPUT" ) );
+  addParameter( rings.release() );
+
+  std::unique_ptr< QgsProcessingParameterNumber> distance = qgis::make_unique< QgsProcessingParameterNumber >( QStringLiteral( "DISTANCE" ),
+      QObject::tr( "Distance between rings" ), QgsProcessingParameterNumber::Double,
+      1, false, 0 );
+  distance->setIsDynamic( true );
+  distance->setDynamicPropertyDefinition( QgsPropertyDefinition( QStringLiteral( "DISTANCE" ), QObject::tr( "Distance between rings" ), QgsPropertyDefinition::DoublePositive ) );
+  distance->setDynamicLayerParameterName( QStringLiteral( "INPUT" ) );
+  addParameter( distance.release() );
+}
+
+bool QgsMultiRingBufferAlgorithm::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback * )
+{
+  mRingsNumber = parameterAsInt( parameters, QStringLiteral( "RINGS" ), context );
+  mDynamicRingsNumber = QgsProcessingParameters::isDynamic( parameters, QStringLiteral( "RINGS" ) );
+  if ( mDynamicRingsNumber )
+    mRingsNumberProperty = parameters.value( QStringLiteral( "RINGS" ) ).value< QgsProperty >();
+
+  mDistance = parameterAsDouble( parameters, QStringLiteral( "DISTANCE" ), context );
+  mDynamicDistance = QgsProcessingParameters::isDynamic( parameters, QStringLiteral( "DISTANCE" ) );
+  if ( mDynamicDistance )
+    mDistanceProperty = parameters.value( QStringLiteral( "DISTANCE" ) ).value< QgsProperty >();
+
+  return true;
+}
+
+QgsFields QgsMultiRingBufferAlgorithm::outputFields( const QgsFields &inputFields ) const
+{
+  QgsFields fields = inputFields;
+  fields.append( QgsField( QStringLiteral( "ringId" ), QVariant::Int, QString(), 10, 0 ) );
+  fields.append( QgsField( QStringLiteral( "distance" ), QVariant::Double, QString(), 20, 6 ) );
+  return fields;
+}
+
+QgsFeatureList QgsMultiRingBufferAlgorithm::processFeature( const QgsFeature &feature, QgsProcessingContext &, QgsProcessingFeedback *feedback )
+{
+  double currentDistance = 0;
+  QgsGeometry outputGeometry, previousGeometry;
+
+  QgsFeatureList outputs;
+
+  for ( int i = 1; i <= mRingsNumber; ++i )
```

Comment:
There's a missing bit in here - we need to evaluate the dynamic parameters here for this feature.

So something like:

    int ringsNumber  = mRingsNumber;
    if ( mDynamicRingsNumber )
        ringsNumber = mRingsNumberProperty.valueAsInt( context.expressionContext(), ringsNumber );

    for ( int i = 1; i <= ringsNumber; ++i )

A similar change is required for mDistance - we'll need to evalute that once outside of the loop and then use the evaluated value inside the loop